Multiple delete of datasets/members

katelynienaber opened the pull request:
## Proposed changes

Does what it says on the tin.
- Select a bunch of data sets & members
- Hit your Delete button on the keyboard
- Confirm you want to delete
- The data sets/members you selected are gone
- If you selected favorites or profiles, they won't disappear

## Release Notes

Milestone: Backlog

Changelog: Added multi-delete capability for data sets and members

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [x] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

crawr reviewed (commented):
The delete works well for me with the keyboard option, but I noticed that it did not work when I tried to use the context menu `Delete Data Set` to delete multiple datasets and members. 
Is this as expected ? Can we wrap the existing delete command to also perform multiple deletion?

katelynienaber commented:
@crawr I forgot to check this! It should be fixed now.

JillieBeanSim reviewed (changes requested):
@katelynienaber I am in love with this new functionality!!! I have needed to cleanup my data sets list for some time now but deleting them 1 by 1 seemed so tedious lol.
I do have a simple request about the information pop up where it says `node`, could we change the wording of that to maybe `The following were deleted ...` or something similar?
Also, we have some failing integration tests that deal with deleting data sets that should get fixed with this PR
![image](https://user-images.githubusercontent.com/49491949/120545791-14cac380-c3bd-11eb-8d3b-9c70df360372.png)

I am also seeing the following integration tests fail on this branch and not master so I wonder if the CLI version bump is causing another regression that will need to be addressed in a different `High Priority` issue.
UPDATE/EDIT: I got confused with PRs I was testing, this PR shouldn't have the CLI bump correct? Wonder why the following are failing.
![image](https://user-images.githubusercontent.com/49491949/120546990-7f303380-c3be-11eb-9e98-3119cec75787.png)

katelynienaber commented:
@JillieBeanSim For the deletion text, is `The following items were deleted` acceptable?

Also I removed those 2 failing `deleteDataset` integration tests. There is no more confirmation dialog in function `deleteDataset`, I have moved it into the new function `deleteDatasetPrompt`. That function [already has unit tests]({"profileName":"nienaber","dataSetName":"KN891873.COPY.FROM.SET"}) covering the cancellation dialog.

katelynienaber commented:
@JillieBeanSim I am still getting the 2nd integration test errors, is it a problem for you as well?

JillieBeanSim commented:
@katelynienaber I am still seeing those tests that deal with patterns fail and only on this branch do I see this issue.  These tests directly follow the delete test you have and they use the same testTree so I wonder if the delete is altering the testTree for the following tests.

crawr commented on packages/zowe-explorer/src/dataset/actions.ts:
```diff
@@ -176,6 +176,97 @@ export async function uploadFile(node: ZoweDatasetNode, doc: vscode.TextDocument
+    const quickPickOptions: vscode.QuickPickOptions = {
+        placeHolder: localize(
+            "deleteDatasetPrompt.quickPickOption",
+            "Delete data set(s)? This will permanently remove them from your system."
```
This is just something small that I noticed: it was a bit weird seeing `delete data sets` when I am deleting members. I understand, though, that they share this same quickpick.
It is not a big deal for me since I am working under the data sets tree, and the success info message clearly specifies which members were deleted.

crawr reviewed (approved):
Great work on this @katelynienaber, thanks!

nickImbirev commented on packages/zowe-explorer/src/dataset/actions.ts:
```diff
@@ -191,6 +191,97 @@ export async function uploadFile(node: ZoweDatasetNode, doc: vscode.TextDocument
+                    try {
+                        await deleteDataset(currNode, datasetProvider);
+                    } catch (err) {
+                        // Do nothing
```
Just a suggestion: maybe we can return this error?

So, there will be no super long list of info messages for the user, but one report with success and error items inside.

We can transform this: `await vscode.window.withProgress...` into `const results = await vscode.window.withProgress...` and iterate via the results to show success deletions and error deletions in one message. 

Hope it helps

katelynienaber replied:
All of the succesful items are already shown in one message, and each failed item has its own message with the error. I somewhat like having each error pop up separately...

katelynienaber replied:
Oh, this is a good point. I added a fix, see if it's ok please...

nickImbirev reviewed (commented):
Just a small suggestions

katelynienaber commented on packages/zowe-explorer/src/dataset/DatasetTree.ts:
```diff
@@ -89,7 +89,7 @@ export class DatasetTree extends ZoweTreeProvider implements IZoweTree<IZoweData
         }
         this.mSessionNodes = [this.mFavoriteSession];
-        this.treeView = vscode.window.createTreeView("zowe.explorer", { treeDataProvider: this });
+        this.treeView = vscode.window.createTreeView("zowe.explorer", { treeDataProvider: this, canSelectMany: true });
```
This is a huge change :( I will have to check every command n the menu, and if it doesn't support it, add an error message.

Not saying it doesn't need to be done. It's just a bit late in the game to be requesting it...

nickImbirev replied:
Hmm, but with this change, you are basically create a very bad user experience, when the multiselect is working in the tree but may not working in the commands... Maybe we can check for other commands and create an issue or smt like this to track this effort?

crawr replied:
From what I know, if the context menu does not handle multiple selected nodes, there will be no error messages. It will simply do the action for the actual node you right-clicked on and ignore the other selected nodes.

katelynienaber replied:
Ok, yes, that's correct. According to the [VSCode docs](https://code.visualstudio.com/api/references/vscode-api#TreeViewOptions%3CT%3E), the first argument to any function will be the tree item the command was executed on. So all the old commands will still receive 1 node, where they are expecting it, and should work as usual.

nickImbirev replied:
Yep, I understand it. 
The only risk I see here, is that: users may choose a lot of items, issue some command for them and it will affect only one element (the clicked one). I would be very sad :(

But at the same time, multiselection will work for the deletion of the datasets, so I will be sad and confused :(

katelynienaber commented:
@JillieBeanSim I've fixed the integration tests, please take a look if you have time ^^

JillieBeanSim commented:
@katelynienaber I am still getting the failures on this branch, maybe someone else could run them also and make sure it's not just my.  I know these tests can be finnicky at times.
![image](https://user-images.githubusercontent.com/49491949/121208560-25a98800-c848-11eb-8bd8-55c541f4ae31.png)

jellypuno commented:
I have them as well.

katelynienaber commented:
@JillieBeanSim @jellypuno Could you give me more info about the failures? Because they are passing for me:
![image](https://user-images.githubusercontent.com/58596139/121320589-96ff3e80-c90d-11eb-984d-29576d89a2a6.png)

jellypuno commented:
Can you show me your testProfileData.ts? Mask your credentials

katelynienaber commented:
@jellypuno 
![image](https://user-images.githubusercontent.com/58596139/121322717-8223aa80-c90f-11eb-9f80-8a5aea527f28.png)

jellypuno commented:
@katelynienaber I'm having an error here `const childrenFromTree = await sessionNode.getChildren();` 404 error for CREATE.DATASET

jellypuno commented:
I am running the pre-test and post-test before executing integration

jellypuno commented:
@JillieBeanSim Katelyn and I have been investigating this one and it is a really tricky issue. I think we would need more time to investigate. If it is okay with you, Can we merge this PR and then follow-up on a fix for that integration test?

JillieBeanSim commented:
> @JillieBeanSim Katelyn and I have been investigating this one and it is a really tricky issue. I think we would need more time to investigate. If it is okay with you, Can we merge this PR and then follow-up on a fix for that integration test?

@jellypuno @katelynienaber that should be fine.  I will test functionality again this afternoon and if none of that changed from before throw my approval on it.  Should an issue be created for handling the integration tests or just wing it?

jellypuno commented:
@katelynienaber please open an issue for the integration for tracking.

katelynienaber commented:
@jellypuno @JillieBeanSim [here](https://github.com/zowe/vscode-extension-for-zowe/issues/1354) is the issue for the integration tests.

JillieBeanSim commented:
@katelynienaber deleting  multiple data sets works well with right-click and keyboard , but multiple delete of members doesn't remove the members from the tree until the data set is closed then re-opened.  I see these errors below in the debug console for each member
![image](https://user-images.githubusercontent.com/49491949/121517133-fff1bf80-c9bc-11eb-8095-a386148330ea.png)

![image](https://user-images.githubusercontent.com/49491949/121517043-e6507800-c9bc-11eb-834c-b889856be1c3.png)

katelynienaber commented:
@JillieBeanSim I added a fix for this, please take a look ^^

JillieBeanSim commented:
Thanks @katelynienaber! All LGTM with the update to handle the members being removed from the tree as well during the deletion.  This multi-delete action is so awesome!! Thanks for getting this done for us

crawr reviewed (approved):
LGTM. Thanks @katelynienaber
